Refactor request building

elskwid opened the pull request:
Refactors the request building to use a series of builder classes - one
for each payment method type. Each builder inherits from
`RequestBuilder` and implements only those methods that it supports. For
instance, the `AuthorizationRequestBuilder` (used for any actions called
on an `Authorization` object) only supports `capture`, `refund` , and
`void`, while `CheckRequestBuilder` supports `purchase`, `refund`, and
`store`.

The gateway sets up a small registry of builders during initialization
and those builders are delegated to inside each gateway action. The
requests are built and sent to Vantiv in the normal way.

This elaborate construction is a by product of the way in which Vantiv
expects the XML to be formed for the different kinds of requests. It
makes it explicit and easy to find where a chose action/method is
implemented.

Common xml helpers are in the `RequestBuilderHelpers` module for
organization. Common functionality used by the gateway is implemented in
private methods on the gateway itself.

Commit:
Vantiv: Refactor request building

Refactors the request building to use a series of builder classes - one
for each payment method type. Each builder inherits from
`RequestBuilder` and implements only those methods that it supports. For
instance, the `AuthorizationRequestBuilder` (used for any actions called
on an `Authorization` object) only supports `capture`, `refund` , and
`void`, while `CheckRequestBuilder` supports `purchase`, `refund`, and
`store`.

The gateway sets up a small registry of builders during initialization
and those builders are delegated to inside each gateway action. The
requests are built and sent to Vantiv in the normal way.

This elaborate construction is a by product of the way in which Vantiv
expects the XML to be formed for the different kinds of requests. It
makes it explicit and easy to find where a chose action/method is
implemented.

Common xml helpers are in the `RequestBuilderHelpers` module for
organization. Common functionality used by the gateway is implemented in
private methods on the gateway itself.

smoil reviewed (approved):
This came together nicely!

grantbdev committed:
Vantiv: Break up `add_card_and_source`

- Break up the `add_card_and_source` method because the Vantiv XML
  schema is order-dependent and they must be added to the requests
  in different places.

grantbdev committed:
Vantiv: Update gateway for certification tests

- Add `build_request` keyword argument to disable adding `orderId` nodes
  since they are not allowed in `AuthorizationRequestBuilder` requests.
- Add missing `refund_type` method and change the refund request type
  in `AuthorizationRequestBuilder` if it is for an eCheck.
- Fix the order of request nodes to match documentation.

grantbdev committed:
Vantiv: Rename `add_*` methods to match node names

grantbdev committed:
Vantiv: Add ability to skip non-repeatable tests

- Add an environment variable that can be used to skip non-repeatable
  assertions in the certification tests.
